udsoncan/services/Authentication.py

```diff
@@ -0,0 +1,383 @@
+           The value is left aligned and right padded with zero up to 16 bytes.
+        :type algorithm_indicator: bytes
+
+        :param certificate_evaluation_id: Optional unique ID to identify the evaluation type of the transmitted
+         certificate. The value of this parameter is vehicle manufacturer specific. Subsequent diagnostic requests with
+          the same evaluationTypeId will overwrite the certificate data of the previous requests.
+           Allowed values are from 0 to 0xFFFF.
+        :type certificate_evaluation_id: int
+
+        :param certificate_data: Optional The Certificate to verify.
+        :type certificate_data: bytes
+
+        :param proof_of_ownership_client: Optional Proof of Ownership of the previous given challenge to be verified by
+         the server.
+        :type proof_of_ownership_client: bytes
+
+        :param ephemeral_public_key_client: Optional Ephemeral public key generated by the client for Diffie-Hellman key
+         agreement.
+        :type ephemeral_public_key_client: bytes
+
+        :param additional_parameter: Optional additional parameter is provided to the server if the server indicates
+         as neededAdditionalParameter.
+        :type additional_parameter: bytes
+
+        :raises ValueError: If parameters are out of range, missing or wrong type
+        """
+
+        tools.validate_int(authentication_task, min=0, max=8, name='Authentication Task')
+        data = None
+        if authentication_task in (Authentication.AuthenticationTask.deAuthenticate,
+                                   Authentication.AuthenticationTask.authenticationConfiguration):
+            # Service ID + Subfunction, no param
+            pass
+        elif authentication_task in (Authentication.AuthenticationTask.verifyCertificateUnidirectional,
+                                     Authentication.AuthenticationTask.verifyCertificateBidirectional,
+                                     Authentication.AuthenticationTask.requestChallengeForAuthentication):
+            tools.validate_int(communication_configuration, min=0, max=0xFF, name='Communication Configuration')
+            # communicationConfiguration
+            data = struct.pack('B', communication_configuration)
+            if authentication_task in (Authentication.AuthenticationTask.verifyCertificateUnidirectional,
+                                       Authentication.AuthenticationTask.verifyCertificateBidirectional):
+                # certificateClient
+                data = Authentication._append_byes_parameter(data, certificate_client, 'Certificate Client')
+
+                # challengeClient
+                data = Authentication._append_byes_parameter(data, challenge_client, 'Challenge Client')
+            else:
+                if not isinstance(algorithm_indicator, bytes) or len(algorithm_indicator) != 16:
+                    raise ValueError(f'{algorithm_indicator} must be a bytes object of length 16')
```

Comment:
Can you make mypy happy about this format string?

`udsoncan/services/Authentication.py:225: error: If x = b'abc' then f"{x}" or "{}".format(x) produces "b'abc'", not "abc". If this is desired behavior, use f"{x!r}" or "{!r}".format(x). Otherwise, decode the bytes`

Reply:
will fix